src/metabase/upload.clj

```diff
@@ -282,25 +281,23 @@
 (defn- file-mime-type [^File file]
   (.detect tika file))
 
-(defn- detect-charset ^String [file]
-  (try
-    (let [detector (UniversalDetector.)
-          buffer   (byte-array 8192)]
-      (with-open [input-stream (io/input-stream file)]
-        (loop []
-          (let [bytes-read (.read input-stream buffer)]
-            (if (pos? bytes-read)
-              (do
-                (.handleData detector buffer 0 bytes-read)
-                (when-not (.isDone detector)
-                  (recur)))
-              (.dataEnd detector)))))
-      (.getDetectedCharset detector))
-    (catch Exception _)))
+(def ^:private supported-charsets
+  #{"UTF-8"
+    "UTF-16" "UTF-16BE" "UTF-16LE"
+    "UTF-32" "UTF-32BE" "UTF-32LE"
+    "WINDOWS-1252"})
+
+(defn- detect-charset ^String [^File file]
+  (or
+   (try
+     ;; If its not a first-class supported encoding, just treat it as the default encoding.
+     (supported-charsets (UniversalDetector/detectCharset file))
```

Comment:
It turns out that they already had a helper for the tedious stream reading code. Basically the same as ours, with a smaller buffer.